drivers/adc: add Gecko IADC driver

fkokosinski opened the pull request:
This PR adds the Gecko IADC driver and support for it to the efr32bg_sltb010a board.

msierszulski committed:
drivers/adc: add Gecko IADC driver

This commit adds the Gecko IADC driver and support for it to the
efr32bg_sltb010a board.

Signed-off-by: Mateusz Sierszulski <msierszulski@antmicro.com>
Signed-off-by: Roman Dobrodii <rdobrodii@antmicro.com>

fkokosinski commented:
Hi @anangl! Could take a look at this PR?

kgugala reviewed (dismissed):
LGTM

anangl commented on drivers/adc/iadc_gecko.c:
```diff
@@ -0,0 +1,482 @@
+		return res;
+	}
+
+	adc_gecko_set_config(data->dev);
```
Why is it called here when `data->channel_id` has not been set yet?

fkokosinski replied:
You're right that it shouldn't be called here. Removed.

fkokosinski commented on drivers/adc/iadc_gecko.c:
```diff
@@ -0,0 +1,482 @@
+	if (res < 0) {
+		return -EINVAL;
+	}
+	data->resolution = res;
```
Thanks for spotting it. It indeed isn't used, so I've removed it.

fkokosinski commented on dts/bindings/adc/silabs,gecko-iadc.yaml:
```diff
@@ -0,0 +1,15 @@
+    required: true
+
+  interrupts:
+    required: true
```
Done

fkokosinski commented on drivers/adc/iadc_gecko.c:
```diff
@@ -0,0 +1,482 @@
+	data->resolution = res;
+
+	/* Verify all requested channels are initialized and store resolution */
+	channels = sequence->channels;
```
Introduced a check here.

fkokosinski commented:
@anangl thanks for the review. Please let me know if anything else needs addressing/fixing. The reported regression (twister job failure) is not introduced by this PR.

fkokosinski committed:
tests/drivers/adc/adc_api/boards: add `efr32bg22_brd4184a` overlay

This commit adds support for the `drivers.adc` test by adding an overlay
for the `efr32bg22_brd4184a` board.

Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

fkokosinski commented:
Hi @anangl! All checks except the DNM one are now green after https://github.com/zephyrproject-rtos/zephyr/pull/56806 got merged.

fkokosinski commented:
Hi @anangl! I had to rebase the PR due to Zephyr and HAL changes. Could you PTAL?
